aten/src/ATen/native/ConvUtils.h

```diff
@@ -214,4 +218,179 @@ static inline bool miopen_conv_use_channels_last(const at::Tensor& input, const
+  } else {
+    return weight;
+  }
+}
+
+// Computes finput used in slow 2D kernel computation.
+// This is computed separately in the forward and backward passes.
+static Tensor compute_finput2d(
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding) {
+  const Tensor weight_2d = view_weight_2d(weight);
+  const int64_t kernel_height = kernel_size[0];
+  const int64_t kernel_width = kernel_size[1];
+  const int64_t pad_height = padding[0];
+  const int64_t pad_width = padding[1];
+  const int64_t stride_height = stride[0];
+  const int64_t stride_width = stride[1];
+  const int64_t dim_planes = 1;
+  const int64_t dim_height = 2;
+  const int64_t dim_width = 3;
+  const int64_t n_input_plane = input.size(dim_planes);
+  const int64_t input_height = input.size(dim_height);
+  const int64_t input_width = input.size(dim_width);
+  const int64_t n_output_plane = weight_2d.size(0);
+  const int64_t output_height =
+      (input_height + 2 * pad_height - kernel_height) / stride_height + 1;
+  const int64_t output_width =
+      (input_width + 2 * pad_width - kernel_width) / stride_width + 1;
+  const int64_t batch_size = input.size(0);
+
+  Tensor finput = at::empty({0}, input.options());
+  if ((input.ndimension() == 4) && (kernel_height == 1) && (stride_height == 1) && (pad_height == 0) &&
+      (kernel_width == 1) && (stride_width == 1) && (pad_width == 0)) {
+    finput = input.view(
+      {batch_size,
+      n_input_plane,
+      output_height * output_width}).detach();
+  } else {
+    finput.resize_(
+      {batch_size,
+        n_input_plane * kernel_height * kernel_width,
+        output_height * output_width});
+
+    AT_DISPATCH_ALL_TYPES_AND(kBFloat16, input.scalar_type(), "compute_finput2d", [&]{
+      auto input_a = input.accessor<scalar_t, 4>();
+      auto finput_a = finput.accessor<scalar_t, 3>();
+
+      at::parallel_for(0, batch_size, 0, [&](int64_t start, int64_t end) {
+        for (int64_t t = start; t < end; t++) {
+          auto input_t = input_a[t];
+          auto finput_t = finput_a[t];
+          unfolded2d_copy_stub(
+            kCPU,
+            c10::CppTypeToScalarType<scalar_t>::value,
+            finput_t.data(),
+            input_t.data(),
+            kernel_height,
+            kernel_width,
+            stride_height,
+            stride_width,
+            pad_height,
+            pad_width,
+            n_input_plane,
+            input_height,
+            input_width,
+            output_height,
+            output_width);
+        }
+      });
+    });
+  }
+
+  return finput;
+}
+
+// Computes finput used in slow 3D kernel computation.
+// This is computed separately in the forward and backward passes.
+static Tensor compute_finput3d(
+    const Tensor& input_,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding) {
+  const int64_t kernel_depth = kernel_size[0];
```

Comment:
I don't see any deletions, where did you get `compute_finput2d/compute_finput3d` from? Is there a summary on what exactly `finput2d / finput3d` are?

Reply:
That's a great question - I got this from the forward pass logic, which computes `finput` alongside the other calculations. For better or worse, I left the forward pass entirely alone and just copied the finput-specific parts out.

It's honestly still not entirely clear to me what exactly `finput` is - let me see if I can figure this out. Maybe there's a way to sidestep the need for it without too much of a rewrite of the slow kernels.

Reply:
I'm wondering if it's possible to refactor out the finput logic so that it can be used by both the convolution_backward and the forward pass logic to avoid code duplication

Reply:
Update: `finput` is a columnized form of the input used for matmul with the flattened kernels. So the slow algos essentially do unfold -> matmul -> fold.

I simplified this quite a bit by just calling `im2col` for the 2D case. Unfortunately, that function doesn't work for the 3D case so it's still calling the `Unfold3dCopyCPU`. Interestingly, there's no CUDA kernel for the slow 3D case (probably because the CUDA kernel depends on `im2col` and it doesn't work for 3D.)